stlab/concurrency/utility.hpp

```diff
@@ -108,17 +118,60 @@ T blocking_get(future<T> x) {
     return std::move(result.get());
 }
 
+template <typename T>
+boost::optional<T> blocking_get(future<T> x, const std::chrono::nanoseconds& timeout) {
+    boost::optional<T> result;
+    std::exception_ptr error = nullptr;
+    auto state = std::make_shared<detail::shared_state>();
+    auto hold = std::move(x).recover(immediate_executor, [_weak_state = make_weak_ptr(state), &result, &error](auto&& r) {
+        auto state = _weak_state.lock();
+        if (!state) {
+            return;
+        }
+        if (r.error())
+            error = std::forward<decltype(r)>(r).error().value();
+        else
+            result = std::forward<decltype(r)>(r).get_try().value();
```

Comment:
The access to result must happen under the lock. Otherwise you have a race in the moment of a timeout.
As well it can happen, that you just entered the recover function and are beyond the `if (!state)` and then the timeout elapses and the function is left, in this case the access to error and result is not very wise.
You have always to keep in mind, that a thread may be suspended at any point in time.

Reply:
But of course. Will move everything into the shared state.